Source/Core/Core/AchievementManager.h

```diff
@@ -0,0 +1,47 @@
+// Copyright 2023 Dolphin Emulator Project
+// SPDX-License-Identifier: GPL-2.0-or-later
+
+#pragma once
+
+#ifdef USE_RETRO_ACHIEVEMENTS
+#include <mutex>
+#include <string>
+#include <thread>
+
+#include <rcheevos/include/rc_api_user.h>
+#include <rcheevos/include/rc_runtime.h>
+
+#include "Common/Event.h"
+
+class AchievementManager
+{
+public:
+  static AchievementManager* GetInstance();
+  void Init();
+  void RequestPost();
+  void SetPassword(const std::string& password);
+
+private:
+  AchievementManager() = default;
+
+  void ThreadProc();
+
+  void Login(const std::string& password);
+
+  void Logout();
+
+  template <typename RcRequest, typename RcResponse>
```

Comment:
Can you add some documentation about why this is made generic via templates? From the point of view of someone reading the code with no or little knowledge of the rcheevos API this looks extremely odd, so the code should be made more self-explanatory.